classes/migration/upgrade/v3_5_0/I7135_CreateNewRorRegistryCacheTables.php

```diff
@@ -0,0 +1,171 @@
+            $table->bigIncrements('ror_setting_id');
+            $table->bigInteger('ror_id');
+            $table->string('locale', 28)->default('');
+            $table->string('setting_name', 255);
+            $table->mediumText('setting_value')->nullable();
+
+            $table->foreign('ror_id')
+                ->references('ror_id')->on('rors')->cascadeOnDelete();
+            $table->unique(['ror_id', 'locale', 'setting_name'], 'ror_settings_unique');
+        });
+
+        // update the tables with latest data set dump from Ror.org
+        $updateRorRegistryDataset = new UpdateRorRegistryDataset();
+        $updateRorRegistryDataset->execute();
+
+        $this->migrateRorAffiliations();
+
+        $this->migrateNonRorAffiliations();
+    }
+
+    /**
+     * Reverse the downgrades
+     *
+     * @throws DowngradeNotSupportedException
+     */
+    public function down(): void
+    {
+        throw new DowngradeNotSupportedException();
+    }
+
+    /**
+     * Migrate affiliations with an exact name in rors table.
+     * Only migrate if author has none in author_affiliations.
+     *
+     * select distinct r.ror_id, a_s.author_id
+     * from rors as r
+     * join ror_settings as r_s on r.ror_id = r_s.ror_id
+     * join author_settings as a_s on r_s.setting_value = a_s.setting_value
+     * where r_s.setting_name = 'name' and a_s.setting_name = 'affiliation' and r_s.locale = a_s.locale
+     *
+     * @return void
+     */
+    public function migrateRorAffiliations(): void
+    {
+        $rows = DB::table('rors as r')
+            ->join('ror_settings as r_s',
+                function (JoinClause $join) {
+                    $join
+                        ->on('r.ror_id', '=', 'r_s.ror_id')
+                        ->where('r_s.setting_name', '=', 'name');
+                }
+            )
+            ->join('author_settings as a_s',
+                function (JoinClause $join) {
+                    $join
+                        ->on('r_s.setting_value', '=', 'a_s.setting_value')
+                        ->on('r_s.locale', '=', 'a_s.locale')
+                        ->where('a_s.setting_name', '=', 'affiliation');
+                }
+            )
+            ->leftJoin('author_affiliations as aa', 'a_s.author_id', '=', 'aa.author_id')
+            ->where('aa.author_id', '=', null)
+            ->select(['r.ror_id', 'a_s.author_id'])
+            ->distinct()
+            ->get();
+
+        foreach ($rows as $row) {
+            $ror = Repo::ror()->get($row->ror_id);
+
+            $affiliation = Repo::affiliation()->newDataObject();
+            $params = [
+                "id" => null,
+                "authorId" => $row->author_id,
+                "ror" => $ror->_data['ror'],
+                "name" => $ror->_data['name']
+            ];
+            $affiliation->setAllData($params);
+
+            Repo::affiliation()->dao->updateOrInsert($affiliation);
```

Comment:
In upgrade scripts we are using only SQLs and no objects and functions -- so that the upgrade can be correctly done for the given version even if we change the objects and code.
Thus, we should try to use the query builder to insert or update the affiliations.